mautic/mautic
# Fix form rendering/thrown exception for Focus items that use the editor/HTML content modes

[//]: # ( Please answer the following questions: )

| Q  | A
| --- | ---
| Bug fix? | y
| New feature? | 
| Related user documentation PR URL | 
| Related developer documentation PR URL | 
| Issues addressed (#s or URLs) | 
| BC breaks? | 
| Deprecations? | 

[//]: # ( Note that all new features should have a related user and/or developer documentation PR in their respective repositories. )

[//]: # ( Required: )
#### Description:

1. When saving a Focus item with custom HTML, an exception is thrown due to trying to generate contact tokens. This was fixed by checking for a lead before attempting to obtain data about the lead.
2. The contact tokens were generated at the time the content was cached so it was not live. This was fixed by replacing tokens at the time the focus content is fetched.
3. Forms manually embedded for the editor/HTML modes would not be caught as a conversion. This was fixed by using a new {focus_form} token that will inject the content of form selected. That way the custom rendering for the form required to make the focus conversion work is generated. 
4. If the HTML/editor content has images, the first time the iframe loads may be the wrong size due to the browser not having fetched the image yet. This was fixed by resizing within the iframe's onload method. 
5. The bar was broken for basic content as the editor/HTML content always rendered.

[//]: # ( As applicable: )
#### Steps to reproduce the bug:
1. Create a new focus item, choose HTML for content mode, and attempt to save. It'll kill the buttons.
2. This is not possible to reproduce due to the first bug.
3. The only way to embed a form into the focus item using editor/HTML mode was to copy in the form's generate.js or iframe code. Do so and add the focus item to a webpage and load. Submit the form. Note that it will not show up as a conversion. 
4. Add an large image to the custom content of a focus item. Save and embed the focus item into a web page. Use a guest or incognito browser to load the page for the first time. The content will not be fully visible. Refresh and it will be. 
5. Create a new focus item and select a emphasize link focus and a bar style. Then choose basic for content and notice that the bar will have no content.

#### Steps to test this PR:
1. Save and should work
2. Add a contact token, ensure you are tracked by that token in a Guest session, open a focus item and the content should have the contact name. 
3. When building the focus item, choose the editor (test) then HTML (test), choose the form then insert {focus_form} into the content. The form's content should load in the preview and in the production focus item.
4. The focus should show the entire content. 
5. Repeat and this time the basic content will show (button) or if you choose editor/HTML, the appropriate content will show.
